### PR TITLE
Filter to expressed variants

### PR DESCRIPTION
**Pull request by tavinathanson**
In this PR:
- Factored some of the `df_isovar` logic out of `load_neoantigens` into a separate `_load_single_patient_isovar`, to use it outside the neoantigen context.
- Added `_filter_variants_to_expressed` to filter out any variant that isn't present in the `isovar` `DataFrame`.
- Randomly made it into this branch: parameterizing `variant_qc_filter`.

@arahuja

### CODE REVIEW

File: cohorts/load.py [hunk 1]
```diff
@@ -748,13 +800,52 @@ def load_neoantigens(self, patients=None, variant_type="snv", merge_type="union"
                 dfs[patient.id] = df_epitopes
         return dfs
 
+    def _load_single_patient_isovar(self, patient, variant_type, merge_type, epitope_lengths):
```
Comment (tavinathanson) [inline cohorts/load.py]: I believe the content of this is all the same; just moved.

File: cohorts/load.py [hunk 2]
```diff
@@ -533,8 +547,31 @@ def _load_single_patient_variants(self, patient, variant_type, merge_type, filte
...
+    def _filter_variants_to_expressed(self, variants, patient, variant_type, merge_type):
+        """Given a VariantCollection, filters that collection to only expressed variants."""
+        if len(variants) == 0:
+            return variants
+
+        expressed_variants = []
+        expressed_metadata = {}
+        # TODO: we're currently using the same isovar cache that we use for expressed
+        # neoantigen prediction; so we pass in the same epitope lengths.
+        # This is hacky and should be addressed.
```
Comment (tavinathanson) [inline cohorts/load.py]: Open to ideas; wasn't sure how to best clean this up.

File: cohorts/load.py [hunk 3]
```diff
@@ -533,8 +547,34 @@ def _load_single_patient_variants(self, patient, variant_type, merge_type, filte
...
+                                                     variant_type=variant_type,
+                                                     merge_type=merge_type)
+        for i, row in df_isovar.iterrows():
+            genome = variants[0].ensembl
+            variant = Variant(contig=row["chr"],
+                              start=row["pos"],
+                              ref=row["ref"],
+                              alt=row["alt"],
+                              ensembl=genome)
+            if variant in variants:
```
Comment (arahuja) [inline cohorts/load.py]: This is going to go through all of `variants` each time, no?

File: cohorts/load.py [hunk 4]
```diff
@@ -652,16 +697,26 @@ def load_effects(self, patients=None, only_nonsynonymous=False, variant_type="sn
...
+                                     variant_type, merge_type,
+                                     only_expressed, filter_fn):
+        def on_return(variants, effects):
+            if only_expressed:
+                variants = self._filter_variants_to_expressed(variants,
+                                                              patient=patient,
+                                                              variant_type=variant_type,
+                                                              merge_type=merge_type)
+                effects = [effect for effect in effects if effect.variant in variants]
+            return filter_effects_with_metadata(effects, variants.metadata, filter_fn)
```
Comment (arahuja) [inline cohorts/load.py]: This works, but the first arg of `filter_effects_with_metadata` was supposed to be an `EffectCollection`

File: cohorts/load.py (hunk 3 above)
Comment (tavinathanson) [inline cohorts/load.py]: As opposed to caching the output? Yeah, I wanted to try to simplify by not caching this.

File: cohorts/variant_filters.py [hunk 5]
```diff
@@ -20,26 +20,32 @@
 import pandas as pd
 from os import path
 
-def variant_qc_filter(variant, variant_metadata):
+DEFAULT_DEPTH = 30
```
Comment (arahuja) [inline cohorts/variant_filters.py]: Cool!

File: cohorts/load.py (hunk 3 above)
Comment (tavinathanson) [inline cohorts/load.py]: Oh, maybe you mean that I'm iterating through the `VariantCollection` on each check; good point.

**Comment by arahuja** [comment]
My thoughts on this are I'm not sure it's worth building in the isovar filtering as special case to the `load_*` functions and leaving it as something to do after the fact or making a `filter_fn` that could handle it.

**Comment by tavinathanson** [comment]
@arahuja curious why not?

File: cohorts/load.py (hunk 4 above)
Comment (tavinathanson) [inline cohorts/load.py]: Good catch, I blame Python.

**Comment by tavinathanson** [comment]
We spoke offline; @arahuja clarified that the `cohorts.varcode_utils` functions are more general than I realized, so I can indeed plop this logic in there.

**Comment by tavinathanson** [comment]
@arahuja I'm not entirely clear on the best way to do this, since my `variant_expressed_filter` and `effect_expressed_filter` need more than `variant`/`effect` and `variant_metadata` (the `cohort`, `patient`, `variant_type`, `merge_type`). I could put `variant_expressed_filter` inside a count function and make use of the closure, but that's not easily passed to `load_variants`.

Perhaps I should replace `variant_metadata` with heavier-weight object that has the `VariantCollection` metadata as well as `cohort`, `patient`, etc.?

**Comment by arahuja** [comment]
> Perhaps I should replace variant_metadata with heavier-weight object that has the VariantCollection metadata as well as cohort, patient, etc.?

Perhaps? I think needing cohort would be odd, but it somewhat shows that some of the `load_single` functions should probably be pushed into `Patient`

**Comment by tavinathanson** [comment]
@arahuja I agree re pushing more functions to `Patient`; for another time, though :)

**Comment by tavinathanson** [comment]
12 days later, some updates to this PR:
- Create `FilterableVariant`, `FilterableEffect`, etc. as a way to cleanly specify `filter_fn`s as single-parameter functions. Also removes the need for `neoantigen_qc_filter` vs. `variant_qc_filter` since `FilterableNeoantigen` is a `FilterableVariant`.
- Use the above to handle expression filtering instead of adding `only_expressed` everywhere.
- Fix https://github.com/hammerlab/cohorts/issues/73.
- Fix https://github.com/hammerlab/cohorts/issues/66.
- Move `merge_type` and `variant_type` into `Cohort` so they aren't passed around to a million methods.
- Add additional parameters to `variant_qc_filter`.

Back to you @arahuja

File: cohorts/load.py [hunk 6]
```diff
@@ -215,10 +223,16 @@ def __init__(self,
                  join_how="inner",
                  check_provenance=False,
                  polyphen_dump_path=None,
-                 pageant_coverage_path=None):
+                 pageant_coverage_path=None,
+                 variant_type="snv",
```
Comment (arahuja) [inline cohorts/load.py:231]: any reason this is a cohort level property now?

File: cohorts/load.py [hunk 7]
```diff
@@ -360,7 +378,7 @@ def is_lambda(func):
         if type(on) == FunctionType:
             return apply_func(on, col, df)
 
-        # For multiple functions, don't allow kwargs since we won't know which functions
+        # For multiple functions, don't allow kwargs since we won"t know which functions
```
Comment (arahuja) [inline cohorts/load.py]: weird quote change

File: cohorts/load.py (hunk 6 above)
Comment (tavinathanson) [inline cohorts/load.py:231]: Yeah, it became cumbersome and wordy to pass it around everywhere. Agree it's a little weird--

Comment (arahuja) [inline cohorts/load.py:231]: Yea I think I prefer the alternative, but also see the issue that you can pass one to `load_variants` and  a different set to `load_effects` which could be confusing, so I'm fine with this

File: cohorts/load.py (hunk 7 above)
Comment (tavinathanson) [inline cohorts/load.py]: Fixed

**Pull request merged**